### DISCUSSION TITLE
Amiga Assembly version 1.5.0 (pre-release) - 1.6.0 (stable)

### DISCUSSION THREAD

**Discussion by prb28** [prb28/vscode-amiga-assembly#222] [General]
Feedback on the 1.5.0 pre-release version.

**Comment by prb28** [comment]
- [Feat #221](https://github.com/prb28/vscode-amiga-assembly/pull/221) *grahambates* Use of external debugger
    - Conditional breakpoints
    - Log points
    - Custom registers in variables list, including named sub-fields for specific bit ranges
    - Additional formats for variables, including signed and size variants
    - Label offset annotations for address registers

**Reply by prb28** [reply]
@grahambates pre-release done

**Comment by Real-MJoe** [comment]
The new features are just awesome! Is there a small HowTo about the new features e.g. 
- Use of external debugger
- Conditional breakpoints

thx in advance

**Reply by grahambates** [reply]
Here's how to set a conditional breakpoint:

Right click on the line where you want to set the breakpoint and select 'Add conditional breakpoint...'. You can also create log points from this menu.

<img width="417" alt="image" src="https://user-images.githubusercontent.com/1519709/172623444-6c349f3b-bc14-4ece-bc6e-662db0b33688.png">

Enter an expression to be evaluated to determine whether the breakpoint should apply. This can include register names, symbols and standard operators. Numbers can be decimal (default), hex (`$` or `0x`), binary (`%` or `0b`) or octal (`@` or `0o`).

<img width="405" alt="image" src="https://user-images.githubusercontent.com/1519709/172623799-ca4f97bb-7605-4522-8af2-6d20a6daef0d.png">

Execution will stop when the condition is met.

<img width="613" alt="image" src="https://user-images.githubusercontent.com/1519709/172624106-66d7bd77-2c21-4f40-905f-9d1482334ed5.png">

Use of external debugger refers to the fact that this extension now uses the [uae-dap](uae-dap) package, which brings all of this functionality to editors other than VS Code which support the Debug Adapter Protocol standard. This includes NeoVim (config example in the README), and Emacs, for which @themkat has created this environment config [emacs-m68k](https://github.com/themkat/emacs-m68k).

**Reply by Real-MJoe** [reply]
Many many thanks.

**Comment by grahambates** [comment]
@prb28 Here are a few thoughts on features I'd really like to implement but are likely to need work on the emulator side.

- Data breakpoints for registers. We currently have memory breakpoints (for WinUAE) but for me what would be really useful is breaking when a register itself is modified, rather than the memory it points to. I haven't really looked into this yet but I'm guessing it's non-trivial.
- Access to CIA and other hardware addresses. It's not possible to read the CIA addresses from memory, as they don't really exist in a memory bank, but rather in internal variables in `cia.c`. The console debugger provides access to these via the `c` command, which calls `dumpcia()`, but we have no equivalent. I'm thinking the cleanest way to do this might be to treat them as registers, and allow access via a special index value in the same way that `REGISTER_COPPER_ADDR_INDEX` works for the copper address.
- Enabling visual DMA debugging. This is another thing that I always need to resort to the console debugger to access.

Another non debugger specific thing I was thinking about is adding a visual preview of sprites, inspired by [DeZog](https://github.com/maziac/DeZog). Potentially this could be implemented for bitplanes and color indexes too.

**Reply by prb28** [reply]
The emulator side is kind of a complex problem.
The fsuae version is too old and needs to be updated to the last version. I don't know if it is interesting ton invest a lot of time in this old version. But since the debugger code has never been merged to the winuae main branch, the update must be done externally to the corresponding fsuae version.
Moreover the fsuae build is tricky. I don't know it it is easier in the latest release.
I have some hope integrating the vAmiga emulator. It will bring a new multi platform emulator more easy to build and debug. But it's a work in progress. And I don't have much time lately, so it's a slow process.
To answer your features ideas. They are all good, some may be tricky, I've tried to stay as close as I can to the default implementation of the gdb debug protocol to stay compatible with other debuggers.
For winuae, it's important to stay close to the bartman implementation to be compatible with his debugger and use his improvements whenever he adds some.
Data breakpoints for registers, maybe it's easier in the winuae branch. The question is how to implement it in the protocol ? Maybe With a false address memory ? (I don't like this idea🤷‍♂️)
Cia as registers, why not, but it mis be implemented with the fpu registers to keep this natural order.
Dma debugging: maybe bartman did it in winuae? He did some improvement lately, maybe a merge can bring some new features already implemented.
Sprite/bitmap visualization: this one does not need the native code modification and seems interesting too.
I don't do a lot of asm coding, so you have a better vision of what is the best feature.

**Reply by grahambates** [reply]
Just thinking about the best way to address the fs-uae vs WinUAE issue. It doesn't look like it's such a big deal to get the fs-uae changes working against the current version. I'm just having a look at this now and I think I'm pretty close to having it working. However it would be much better to have a single implementation the could potentially be merged upstream.

My understanding is that the fs-uae maintainer was of the opinion that any solution should be applied upstream in WinUAE. In theory bartman's solution looks like a better candidate for this and seems much cleaner in general. However the GDB server part of it is currently using win32 APIs. I'm wondering if the way forward would be to make this cross platform so that it could be incorporated into fs-uae too. I can see that there are GUI features that would have to remain Windows specific, but perhaps it's just a matter of decoupling these. I'll spend some time looking into it.

The vAmiga integration sounds great too and I wonder if [vAmigaWeb](https://github.com/vAmigaWeb/vAmigaWeb) has the potential to allow running the emulator inside VS Code.

**Reply by phaze101** [reply]
@grahambates I was hoping you suggest the integration of vAmigaWeb in the Visual code. Looking forward to such a thing.

@grahambates Log points what are they? I didn't get this.

**Reply by grahambates** [reply]
[Logpoints](https://code.visualstudio.com/docs/editor/debugging#_logpoints) are like breakpoints, but they output text instead of stopping the program, a bit like doing print debugging using `console.log` etc. The text can include expressions:

<img width="411" alt="image" src="https://user-images.githubusercontent.com/1519709/173182827-780f586e-fe66-4f39-8b02-c9c1092cb27a.png">

You can see the output in the debug console:

<img width="620" alt="image" src="https://user-images.githubusercontent.com/1519709/173182856-11f83c56-d5d5-44c7-b599-275ff14efa14.png">

**Reply by prb28** [reply]
I agree on vAmiga, maybe @mithrendal could share his point of view on this (in any language he wants (he likes to mix them ;-))

**Reply by phaze101** [reply]
@grahambates wow thank you. This is awesome.

**Reply by prb28** [reply]
I had a discussion on discord about the winuae/fs-uae integration with bartman, tonioni(Winuae) and Frode(fs-uae). But there isn't a real full motivation because there is already a good console debugging option in these emulators and it's maintained by tonioni directly.

**Reply by grahambates** [reply]
So it looks like GDB support will always be a custom patch/fork then. It would still be nicer if we had a single implementation for both winuae/fs-uae though. I do think it's worth investigating the effort required to make bartman's solution fs-uae compatible.

There is another option, which is to automate interaction with the console debugger like in [cahirwpz's solution](https://github.com/cahirwpz/demoscene/tree/master/tools/debug). This just requires a patch to trigger the debugger on entry but I suspect this doesn't provide all the functionality we require e.g. around stack frames.

**Reply by mithrendal** [reply]
bonjour a tous, hi all,

you can fork and use vAmigaWeb in any context you like 😎 I am a big fan of AmigaAssembly ❤️

At a first glance it seems that vAmigaWeb would be the natural thing and should be quite easy to integrate vAmigaWeb into the VScode extension since both is javascript /HTML5 right?

but vAmigaWeb relies  on some APIs which maybe only exists in real a browser ? I don't know for sure yet...

For example it needs

- WebAudio API
- HTML5 Gamecontroller API
- IndexedDB API
- localstorage API
- WASM
- Service-Worker

these are all supported in all regular WebBrowsers of the last 5 years or so...

a quick google for these APIs in an VSCode extension indicate that WebAudio and WASM should be supported ... for the others it is unclear to me ... one has to try it out...

if one or more APIs are not supported in the VisualStudioCode extension environment then some code in vAmigaWeb has to be patched or removed otherwise it would certainly crash ...

Another topic is the debugger interface ... vAmigaWeb has the same code base as @dirkwhoffmann vAmiga (where already some work to interconnect the two has been done in the past) but that codebase relies on TCP/IP communictation ... this has to be patched ... we don't need a tcp channel when we are in the same Javascript memory space so it is easier and more reliable but for sure there are patches which have to be done for proper communication between amiga assembly and vAmigaWeb...

**Reply by grahambates** [reply]
That sounds really promising. From what I understand Electron (which VS Code is built on) uses the Chromium browser engine and should have feature parity with recent versions of Chrome. VS Code provides [Web Views](https://code.visualstudio.com/api/extension-guides/webview) which are basically like iframes and should give full access to the browser engine.

I'm definitely interested in this approach so I'll do some investigation when I get some time.

**Reply by mithrendal** [reply]
Ah that sounds great. I did not know that. Basically this would also open the door for a still deeper integration as we don’t have to rely on gdb protocol at all, right? the emulator instance would be fully controllable and accessible in any aspect even information gathering stuff beyond the gdb protocol. Cool 😎

**Comment by prb28** [comment]
Wiki updated:
- https://github.com/prb28/vscode-amiga-assembly/wiki/Debugging-emulator#add-a-conditional-breakpoint
- https://github.com/prb28/vscode-amiga-assembly/wiki/Debugging-emulator#add-a-logpoint

**Reply by grahambates** [reply]
I think we could do with some proper documentation around the supported syntax in expressions. I'll put something together.

**Reply by phaze101** [reply]
- Custom registers in variables list, including named sub-fields for specific bit ranges
- Additional formats for variables, including signed and size variants
- Label offset annotations for address registers

Can I suggest that the above should also be documented or commented exactly what they mean to be clear for everyone
.

**Reply by grahambates** [reply]
Yeah I agree these at least need to go in release notes. Just quickly to summarize here:

### Custom registers in variables:
We now list the values of the Amiga hardware registers in the variables panel. It splits them into fields to describe each bit-range and group related registers where appropriate.

<img width="345" alt="image" src="https://user-images.githubusercontent.com/1519709/173191330-47b37f21-1e19-4300-b641-0fe4c788741c.png">

### Additional formats for variables, including signed and size variants
This gives you additional options to format variable values, including the ability to use just the word or byte portion and parse it as either signed or unsigned.

<img width="418" alt="image" src="https://user-images.githubusercontent.com/1519709/173191432-4775975b-7326-463d-a513-ec2d20437fa1.png">
results in:
<img width="348" alt="image" src="https://user-images.githubusercontent.com/1519709/173191440-ba1ea14a-49bc-46de-8d7d-9190799c58af.png">

### Label offset annotations for address registers

This displays the offset from a label in your source code alongside address registers where possible.

<img width="325" alt="image" src="https://user-images.githubusercontent.com/1519709/173191627-eff42b61-2dbf-424e-9a7b-95eb4f6a94d0.png">

**Reply by phaze101** [reply]
Wow this last one is cool. Thanks

**Reply by prb28** [reply]
added here: https://github.com/prb28/vscode-amiga-assembly/wiki/Debugging-emulator#other-debug-features

**Comment by grahambates** [comment]
On reflection I think the size/sign options for variables are better handled as fields in the variables list rather than creating a bucket-load more formatting options in the menu:

<img width="190" alt="image" src="https://user-images.githubusercontent.com/1519709/173197296-161537b5-876b-49d4-b817-ee3e5a3e125d.png">

The big advantage of this is that these can then be used in expressions, for example you could set a conditional breakpoint if `d0.w==0x100`, which would ignore the upper word of the register. This also mirrors how size qualifiers are used with registers in assembly e.g. `(a0,d0.w)`.

I've made the change in uea-dap and will open a PR here shortly.

**Reply by prb28** [reply]
Released in version 1.5.1.

**Comment by grahambates** [comment]
Let me know if there's anything else I can help with to get this release out.

**Reply by prb28** [reply]
I was just waiting to see if there is some feedback from the pre-release.
And harvesting some time to finish.

I still have to rebuild fs-uae on the 3 os and I would like to do it for vasm, vlink and vbcc.

I'll try to do it this week.

**Reply by grahambates** [reply]
Ok no worries. Not bugging you, just checking if there's anything I can do.
